Redefine the call of share and reclaim (#41)
pdidev/tutorial: README.md

`::PDI_share` + `::PDI_event` + `::PDI_reclaim`.

3. `::PDI_multi_expose` for data `A` and data `B` is equivalent to
`::PDI_share` `A` + `::PDI_share` `B`+ `::PDI_event` + `::PDI_reclaim` `B` + `::PDI_reclaim` `A`.

\attention
The `::PDI_multi_expose` is implemented with interlaced share/reclaim pairs.

\attention
When we used `::PDI_multi_expose` with multiple data, the order of appearance
of the arguments of the function corresponds to the order of the `::PDI_share`.

\attention

> In a `::PDI_multi_expose` if you have a data1 that depend on the data2,

Reviewer: ```suggestion
In a `::PDI_multi_expose` if you have a data1 that depends on the data2,
```

\attention
The `::PDI_multi_expose` is implemented with interlaced share/reclaim pairs.

> \attention
> When we used  `::PDI_multi_expose` with multiple data, the order of appearance
> in the arguments of the function corresponds to the order of the `::PDI_share`.

Reviewer: ```suggestion
\attention
When we used `::PDI_multi_expose` with multiple data, the order of appearance
of the arguments of the function corresponds to the order of the `::PDI_share`.
```

> \attention
> To gain a better understanding of the information provided by the trace plugin,
> the order of calls of differents "share" is changed in the ex6
> after the iteration loops.

Reviewer: ```suggestion
\attention
To gain a better understanding of the information provided by the trace plugin,
the call order of different "share" is changed in the ex6
after the main loop.
```

> The order of share are importants when we used a metadata. We explain that
> without `::PDI_multi_expose` to be more clear.
> In `::PDI_share`, the event "on_data" on the shared data are performed.

Reviewer: ```suggestion
The order of `share` is important when using a metadata field. As seen below, 
the exercise 6 decompose a `multi_expose` call to demonstrate this.
Furthermore, when using `::PDI_share`, an event "on_data" 
is automatically performed.
```

An example of this event is
```yaml
decl_hdf5:
    - file: ex6-final-iteration.h5
      write: [ main_field ]
      when: '$ii=4'
```

> In the exercise 6, after the iteration loop, the data `main_field` is shared

Reviewer: ```suggestion
In the exercise 6, after the main loop, the data `main_field` is shared
```

before the variable `ii`:
```C
PDI_share("main_field", cur, PDI_OUT);
PDI_share("ii",         &ii, PDI_OUT); // update the metadata ii in PDI
PDI_event("finalization");
PDI_reclaim("ii");
PDI_reclaim("main_field");
```

> In the first line, `cur` correspond to the value of `main_field` at iteration

Reviewer: ```suggestion
In the first line, `cur` corresponds to the value of `main_field` at iteration
```

`ii=4`. As `ii` is a metadata, the value is stored by pdi.

> Hence, in this first line the value of `ii` is equal to 3.

Reviewer: Ambiguity on index value.

Therefore, the file `ex6-final-iteration.h5` is not writing on the disk.
To solve this issue, we need to change the order of the `::PDI_share`:

```C
PDI_share("ii",         &ii, PDI_OUT); // update the metadata ii in PDI
PDI_share("main_field", cur, PDI_OUT);
PDI_event("finalization");
PDI_reclaim("main_field");
PDI_reclaim("ii");
```

or with `::PDI_multi_expose`:

> ```C
> PDI_multi_expose("finalization",
>           "ii",         &ii, PDI_OUT,
> 	        "main_field", cur, PDI_OUT,
> 	        NULL);
> ```

Reviewer: ```suggestion
```C
PDI_multi_expose("finalization",
	        "ii",         &ii, PDI_OUT,
	        "main_field", cur, PDI_OUT,
	        NULL);
```
```

> \attention
> In a `::PDI_multi_expose` if you have a data1 that depend on the data2.
> You need to pass the data2 before the data1

Reviewer: ```suggestion
\attention
In a `::PDI_multi_expose` if you have a data1 that depends on the data2.
You need to expose the data2 before the data1.
```

> For example, a vector `V` that depend on it size `N`.

Reviewer: ```suggestion
For example, a vector `V` that depends on its size `N`.
```

> ```C
> PDI_multi_expose("save_vector_V",
>           "size_of_vector", &N, PDI_OUT,
> 	        "vector_V", V, PDI_OUT,
> 	        NULL);
> ```

Reviewer: ```suggestion
```C
PDI_multi_expose("save_vector_V",
	        "size_of_vector", &N, PDI_OUT,
	        "vector_V",       V, PDI_OUT,
	        NULL);
```
```
